### ISSUE TITLE
There should be a filter to change max discount amount in get_fees_from_cart function.

### ISSUE THREAD

**Issue by malsubrata** [woocommerce/woocommerce#28028]
**Is your feature request related to a problem? Please describe.**
No this is not a problem but while applying a negative fee in the cart the tax amount is not taking effect.

**Describe the solution you'd like**
There should be a filter to change the max discount amount in the get_fees_from_cart function so that we can add the tax amount to the max_discount variable.

**Describe alternatives you've considered**
alternatively, you can add the cart tax amount to the max_discount variable like 
`$max_discount = round( $this->get_total( 'items_total', true ) + $fee_running_total + $this->get_total( 'shipping_total', true ) + array_sum( $this->get_tax_class_costs() ) ) * -1;`

**Comment by tammullen** [comment]
Hi @malsubrata 

Thank you for taking the time to share this idea, we really appreciate your help. 
​
While we appreciate you sharing your ideas with us, it doesn’t fit in with our current priorities for the project. At some point, we may revisit our priorities and look through the list of suggestions like this one to see if it warrants a second look. In the meantime, we are going to close this issue with the `votes needed` label and evaluate over time if this issue collects more feedback. 
​
Don’t be alarmed if you don’t see any activity on this issue for a while.

**Comment by woocommercebot** [comment]
Thanks for the suggestion @malsubrata.

We'll keep an eye on the [popularity of this request](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md#feature-requests) :)

**Comment by makifujio** [comment]
Hi I hope you solve the problem you @malsubrata as soon as possible, as it is the best wallet plugin. Thanks

**Comment by yeraygomez** [comment]
Hello, do you already have a solution to this? The same thing is happening to me, what is the solution?

Thank you.

**Comment by yeraygomez** [comment]
Hello, this code: $max_discount = round( $this->get_total( 'items_total', true ) + $fee_running_total + $this->get_total( 'shipping_total', true ) + array_sum( $this->get_tax_class_costs() ) ) * -1;

is the solution? in which file should I put it?

**Comment by malsubrata** [comment]
@yeraygomez You can change this code at includes/class-wc-cart-total.php line no 285

**Comment by yeraygomez** [comment]
Hello, we have made the code change you mentioned but the error remains.

It does not calculate the VAT correctly, it still calculates the VAT of the wallet. 

I attach an invoice for you to see: https://tolecarnes.com/wp-admin/admin-ajax.php?action=generate_wpo_wcpdf&document_type=invoice&order_ids=25668&_wpnonce=1f3543e8c6

VAT ("IVA") should be 3,81€.

**Comment by yeraygomez** [comment]
Please! I need a urgent solution... can you please answer me

**Comment by Deardevils** [comment]
It works for me, now the wallet can charge cart total amount which include the tax. 
However this is not an ideal fix as it needs to modify woocommerce core file.
We may need a snippet solution for this.

**Comment by Deardevils** [comment]
> 

It works for me, now the wallet can charge cart total amount which include the tax.
However this is not an ideal fix as it needs to modify woocommerce core file.
We may need a snippet solution for this.

**Comment by Deardevils** [comment]
> @yeraygomez You can change this code at includes/class-wc-cart-total.php line no 285

It works for me, now the wallet can charge cart total amount which include the tax.
However this is not an ideal fix as it needs to modify woocommerce core file.
We may need a snippet solution for this.

**Comment by Poll0none** [comment]
I am here to increase the popularity of this topic and get @malsubrata idea adapted into woocommerce, the wallet plugin he developed is one of the best out there and he is putting a lot of work and dedication into its development and support. 

I can see this plugin getting to be a must for many websites like mine. (Poll0none)

**Comment by ryanaby** [comment]
I'm one of the 20,000+ users who have used Tera Wallet and faced the same issue while setting up a WooCommerce website with tax enabled.

This plugin is one of the best plugins for wallet support, big thanks for @malsubrata for your hard work.

**Comment by abhinavrohatgi** [comment]
I am having the same issue. Link to the issue [https://snipboard.io/nAkQOZ.jpg] 

The cart total is—10,313, and the wallet has a balance of 10,000, but it is not deducting 10,000 and charging the customer 313. The tax rate is 18% for the above 2 items. Instead, it is only allowing to deduct 8740 from the wallet, and charging the customer for the tax amount of 1573.

Kindly help.

**Comment by malsubrata** [comment]
Please we need attention on this. Please add a new filter so that we can modify the amount.

**Comment by asansano** [comment]
@tammullen I think it is one you know to reconsider this highly needed request form @malsubrata . It's just not ok for woocommerce to handling negativ fee taxwise legally wrong. thx for your help on this!

**Comment by fjaka** [comment]
Any fix for this?

**Comment by stheoret** [comment]
Hi great fix, but am i the only one now after this code applied experiencing issues with "cancelled" orders???  The funds are not put back to the wallet of the customer... before i had no issues...

**Comment by dvamvo** [comment]
From what i understand, woccommerce fees are not meant for negative amounts, instead of the negative fee, why dont you use a virtual coupon? or at least give the option in settings so the user has the option for correct tax calculations. @malsubrata 

```php
add_filter('woocommerce_get_shop_coupon_data', 'create_virtual_coupon', 10, 2);

function create_virtual_coupon($data, $code) {
    if ($code !== 'via-wallet') {
        return $data;
    }

    return array(
        'code'           => $code,
        'discount_type'  => 'fixed_cart',
        'amount'         => 2,
        'individual_use' => false,
        'free_shipping'  => false,
    );
}

add_filter('woocommerce_cart_totals_coupon_label', 'my_virtual_coupon_label', 10, 2);

function my_virtual_coupon_label($label, $coupon) {
    if ($coupon->get_code() === 'via-wallet') {
        return 'Via Wallet';
    }

    return $label;
}

add_filter('woocommerce_cart_totals_coupon_html', 'my_virtual_coupon_html', 10, 3);

function my_virtual_coupon_html($coupon_html, $coupon, $discount_amount_html) {
    if ($coupon->get_code() === 'via-wallet') {
        return $discount_amount_html;
    }

    return $coupon_html;
}
